src/ameba/ast/visitors/implicit_return_visitor.cr

```diff
@@ -0,0 +1,216 @@
+module Ameba::AST
+  class ImplicitReturnVisitor < BaseVisitor
+    # When greater than zero, indicates the current node's return value is used
+    @stack : Int32 = 0
+
+    def visit(node : Crystal::Expressions) : Nil
+      @rule.test(@source, node, @stack > 0)
+
+      last_idx = node.expressions.size - 1
+
+      swap_stack do |old_stack|
+        node.expressions.each_with_index do |exp, idx|
+          if exp.is_a?(Crystal::ControlExpression)
+            incr_stack { exp.accept(self) }
+            break
+          elsif idx == last_idx && old_stack > 0
+            incr_stack { exp.accept(self) }
+          else
+            exp.accept(self)
+          end
+        end
+      end
+    end
+
+    def visit(node : Crystal::Call) : Nil
+      @rule.test(@source, node, @stack > 0)
+
+      if node.block
+        incr_stack { node.obj.try &.accept(self) }
+      else
+        node.obj.try &.accept(self)
+      end
+
+      incr_stack {
+        node.args.each &.accept(self)
+        node.named_args.try &.each &.accept(self)
+        node.block_arg.try &.accept(self)
+        node.block.try &.accept(self)
+      }
+    end
+
+    def visit(node : Crystal::Arg) : Nil
+      @rule.test(@source, node, @stack > 0)
+
+      incr_stack { node.default_value.try &.accept(self) }
+    end
+
+    def visit(node : Crystal::EnumDef) : Nil
+      @rule.test(@source, node, @stack > 0)
+
+      node.members.each &.accept(self)
+    end
+
+    def visit(node : Crystal::Assign | Crystal::OpAssign) : Nil
+      @rule.test(@source, node, @stack > 0)
+
+      incr_stack { node.value.accept(self) }
+    end
+
+    def visit(node : Crystal::MultiAssign) : Nil
+      @rule.test(@source, node, @stack > 0)
+
+      node.targets.each &.accept(self)
+      incr_stack { node.values.each &.accept(self) }
+    end
+
+    def visit(node : Crystal::If | Crystal::Unless) : Nil
+      @rule.test(@source, node, @stack > 0)
+
+      incr_stack { node.cond.accept(self) }
+      node.then.accept(self)
+      node.else.accept(self)
+    end
+
+    def visit(node : Crystal::While | Crystal::Until) : Nil
+      @rule.test(@source, node, @stack > 0)
+
+      incr_stack { node.cond.accept(self) }
+      node.body.accept(self)
+    end
+
+    def visit(node : Crystal::Def) : Nil
+      @rule.test(@source, node, @stack > 0)
+
+      incr_stack {
+        node.args.each &.accept(self)
+        node.double_splat.try &.accept(self)
+        node.block_arg.try &.accept(self)
+      }
+
+      return_type = node.return_type
+      case return_type
+      when Crystal::Path
+        # Special case of the return type being nil, meaning the last
+        # line of the method body is ignored
+        if ["::Nil", "Nil"].includes?(return_type.names.join("::"))
```

Comment:
If this is too hacky / "semantic" I can remove it. Just helps catch the edge case of the last line in def bodies being ignored if the return type is nil.